File: webinterface/org.eclipse.rap.addons.d3chart/js/chart/force-directed-graph.js
/*******************************************************************************
 * Copyright (c) 2013 EclipseSource and others. All rights reserved. This
 * program and the accompanying materials are made available under the terms of
 * the Eclipse Public License v1.0 which accompanies this distribution, and is
 * available at http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors: Ralf Sternberg - initial API and implementation
 ******************************************************************************/

d3graph = {};
var that;

d3graph.ForceDirectedGraph = function( parent ) {
    this._width = 900;
    this._height = 700;
    this._color = d3.scale.category20();	
    
    this._renderer = this;
    this._element = this.createElement( parent );
    this._padding = 20;

    this._needsLayout = true;
    that = this;
    
    this._force = d3.layout.force()
	    .charge(-200)
	    .linkDistance(30)
	    .size([this._width, this._height]);
   		
    this._nodes = this._force.nodes()
    this._links = this._force.links()
    
    this._svg = d3.select(this._element).append("svg")
    	.attr("width", this._width)
    	.attr("height", this._height)
    	.attr("class", "chart");
    
    this._svg.append('defs').append('marker')
    .attr({
      id: 'arrowhead',
      refX: 13,
      refY: 4,
      markerUnits: 'strokeWidth',
      markerWidth: 10,
      markerHeight: 8,
      orient: 'auto'
    }).append('path').attr({
      d: 'M 0 0 L 10 4 L 0 8 Z',
      fill: 'black'
    });
    


//    var linkG = this._svg.append("g")
//    var nodeG = this._svg.append("g")
//    this._layer = this.getLayer("layer");    
    this.link_layer = this.getLayer("links")
    this.node_layer = this.getLayer("nodes")
    this.label_layer = this.getLayer("labels")
    
    link = this.link_layer.selectAll(".link")
    		.attr('marker-end', 'url(#arrowhead)');
    node = this.node_layer.selectAll(".node")

	
    rap.on( "render", function() {
      if( that._needsRender ) {
        if( that._needsLayout ) {
          that._renderer.initialize( that );
          that._needsLayout = false;
        }
        that._renderer.render( that );
        that._needsRender = false;
      }
    } );
    
    parent.addListener( "Resize", function() {
      that._resize( parent.getClientArea() );
    } );
    
    this._resize( parent.getClientArea() );
//    this.restart();
};

d3graph.ForceDirectedGraph.prototype = {
		
  addNode: function( node ) {
	this._nodes.push(node);
	this.restart();
  },
  
  addLink: function( link ) {
	this._links.push(link);
	this.restart();
  },
  
  restart: function () {
	var link = this.link_layer.selectAll("line.link")
				.attr('marker-end', 'url(#arrowhead)');
		
	var node = this.node_layer.selectAll(".node");

	var label = this.label_layer.selectAll("text.label");
	
	node.append("title")
      .text(function(d) { return d.name; });
	
	link = link.data(this._links)
	  	.enter()
	  	.append("line")
		.attr("class", "link")
		.attr('marker-end', 'url(#arrowhead)')
		.style("stroke-width", function(d) { return 2.; })
		.style("stroke", function(d) { return "black"; })
		.style("stroke-opacity", function(d) { return Math.max(0.2, d.strength); })
		.style({"z-index":"0"});
		
	link.insert("text")
		.style("stroke", function(d) { return "#000"; })
		.text(function(d) {return d.label})
		.style("stroke-width", .2)
	    .style({"font-family":"Arial, Helvetica, sans-serif","font-size":"12px"}) //,"z-index":"999999999"
	    .style("text-anchor", "middle");
	
	label = label.data(this._links)
	  .enter().append('text')
	  .attr("class", "label")
	    .attr("x", function(d) { return (d.source.y + d.target.y) / 2; }) 
	    .attr("y", function(d) { return (d.source.x + d.target.x) / 2; }) 
	    .attr("text-anchor", "middle") 
	    .text(function(d) {return d.label;});  

	g = node.data(this._nodes)
	  .enter().insert("g")
      .attr("class", "node")
//	  .style({"z-index": "99999"})
    
	g.insert("rect")//.transition().duration(500).attr('width', 150)
//		.attr("class", "node")
      	.style("stroke", "#000")
      	.style("stroke-width", .2)
      	.style({"z-index":"999999"})
      	.style("fill", function(d) { return d.color; })
      	.attr("width", function(d) { return 100;})
      	.attr("height", 20)
		.attr("transform", "translate(-50, -13)")
    	.call(this._force.drag)
//		.enter()
//		.append("circle")
//		.attr("class", "node")
//		.attr("r", 5)
//		.style("fill", "green")
//		.style("stroke", "white");
	
    g.insert("text")
      	.style("stroke", function(d) { return "#000"; })
		.text(function(d) {return d.label})
		.call(this._force.drag)
		.style("stroke-width", .2)
	    .style({"font-family":"Arial, Helvetica, sans-serif","font-size":"12px"}) //,"z-index":"999999999"
	    .style("text-anchor", "middle");

    
 //		.attr("class", "node")
//		.attr("width", 100)
//		.attr("height", 50)
//		.style("fill", function(d) { return "#DC143C"; })
//		.style("stroke-width", function(d) { return "1.5px"; })
//		.call(this._force.drag)
//		.insert("text").text("hello")
	
	link = this.link_layer.selectAll(".link");
	node = this.node_layer.selectAll(".node");
	label = this.label_layer.selectAll(".label");

	  
	this._force.on("tick", function() {
		link.attr("x1", function(d) { return d.source.x; })
			.attr("y1", function(d) { return d.source.y; })
			.attr("x2", function(d) { return d.target.x; })
			.attr("y2", function(d) { return d.target.y; });
//		link.attr("x", function(d) { return Math.min(d.source.x, d.target.x); })
//		.attr("y", function(d) { return Math.min(d.source.y, d.target.y); })
//		.attr("width", function(d) { return Math.max(d.source.x, d.target.x) - Math.min(d.source.x, d.target.x); })
//		.attr("height", function(d) { return Math.max(d.source.y, d.target.y) - Math.min(d.source.y, d.target.y); });
//		link.style({"z-index": "0"})
		node.attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; });
//		node.style({"z-index": "999"})
//		node.attr("cx", function(d) { return d.x; })
//		    .attr("cy", function(d) { return d.y; });
		
//		labels = link.data(this._links)
//	  	.enter().append('text')
//	    .attr("x", function(d) { return (d.source.y + d.target.y) / 2; }) 
//	    .attr("y", function(d) { return (d.source.x + d.target.x) / 2; }) 
//	    .attr("text-anchor", "middle") 
//	    .text(function(d) {return 2;}); 
	
		label.attr("x", function(d) { return (d.source.x + d.target.x) / 2; }) 
        	.attr("y", function(d) { return (d.source.y + d.target.y) / 2; }) 
	});
	
	this._force.start();
  },
  
  createElement: function( parent ) {
    var element = document.createElement( "div" );
    element.style.position = "absolute";
    element.style.left = "0";
    element.style.top = "0";
    element.style.width = "100%";
    element.style.height = "100%";
    parent.append( element );
    return element;
  },

  getLayer: function( name ) {
    var layer = this._svg.select( "g." + name );
    if( layer.empty() ) {
      this._svg.append( "g" ).attr( "class", name );
      layer = this._svg.select( "g." + name );
    }
    return layer;
  },

  _resize: function( clientArea ) {
    this._width = clientArea[ 2 ];
    this._height = clientArea[ 3 ];
    this._force.size([this._width, this._height]);
    this._svg.attr( "width", this._width ).attr( "height", this._height );
    this._scheduleUpdate( true );
  },

  _scheduleUpdate: function( needsLayout ) {
    if( needsLayout ) {
      this._needsLayout = true;
    }
    this._needsRender = true;
  },

  destroy: function() {
    var element = this._element;
    if( element.parentNode ) {
      element.parentNode.removeChild( element );
    }
  },
  
  initialize: function( chart ) {
	    this._chart = chart;
	    this._layer = chart.getLayer( "layer" );
  },
	  
  render: function( chart ) {
	  this.restart();
  },
  
  setCharge: function(charge) {
	  this._force.charge(charge['charge']);
	  this.restart();
  },
  
  setLinkDistance: function(params) {
	  this._force.linkDistance(params['linkdist']);
	  this.restart();
  },
  
  setFriction: function(params) {
	  this._force.friction(params['linkdist']);
	  this.restart();
  },
  
  setChargeDistance: function(params) {
	  this._force.chargeDistance(params['linkdist']);
	  this.restart();
  },
  
  setGravity: function(params) {
	  this._force.gravity(params['linkdist']);
	  this.restart();
  },
  
  removeAllNodes: function(params) {
	  this._nodes.length = 0;
	  this._links.length = 0;
	  this.restart();
  },
};

// TYPE HANDLER

rap.registerTypeHandler( "d3graph.ForceDirectedGraph", {
  factory: function( properties ) {
    var parent = rap.getObject( properties.parent );
    return new d3graph.ForceDirectedGraph( parent );
  },
  
  destructor: "destroy",
 
  methods: ["render", "addNode", "addLink", "restart", "setCharge", "setLinkDistance", "setFriction", 
            "setChargeDistance", "setGravity", "removeAllNodes"],
  
  properties: [],

  events: [ "Selection" ]
  
} );

File: webinterface/org.eclipse.rap.addons.d3chart/src/org/eclipse/rap/addons/d3chart/ForceDirectedGraph.java
/*******************************************************************************
 * Copyright (c) 2014 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ralf Sternberg - initial API and implementation
 ******************************************************************************/
package org.eclipse.rap.addons.d3chart;

import static org.eclipse.rap.rwt.lifecycle.WidgetUtil.getId;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.remote.AbstractOperationHandler;
import org.eclipse.rap.rwt.remote.RemoteObject;
import org.eclipse.swt.SWT;
import org.eclipse.swt.widgets.Canvas;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Event;
import org.eclipse.rap.json.JsonObject;
import org.eclipse.swt.widgets.Item;


public class ForceDirectedGraph extends Canvas {
	
	private static final String REMOTE_TYPE = "d3graph.ForceDirectedGraph";
	private final RemoteObject remoteObject;
	private final List<Item> items = new LinkedList<Item>();
	
	public List<Node> nodes = new ArrayList<Node>();
	public List<Link> links = new ArrayList<Link>();
	
	public ForceDirectedGraph( Composite parent, int style ) {
		    super( parent, style );

		    remoteObject = RWT.getUISession().getConnection().createRemoteObject( REMOTE_TYPE );
		    remoteObject.set( "parent",getId( this ) );		    
		    remoteObject.setHandler( new AbstractOperationHandler() {
		      @Override
		      public void handleNotify( String eventName, JsonObject properties ) {
		        if( "Selection".equals( eventName ) ) {
		          Event event = new Event();
		          event.index = properties.get( "index" ).asInt();
		          event.item = items.get( event.index );
		          notifyListeners( SWT.Selection, event );
		        }
		      }
		    } );
		    
		    ChartResources.ensureJavaScriptResources();
	}
	
	public void addNode(Node node) {
		JsonObject json = new JsonObject();
		json.add("label", node.label);
		json.add("color", node.color);
		node.idx = this.nodes.size();
		remoteObject.call("addNode", json);
		nodes.add(node);
	}
	
	public void addLink(Link link) {
		JsonObject json = new JsonObject();
		json.add("label", link.label);
		json.add("source", link.source.idx);
		json.add("target", link.target.idx);
		json.add("directed", link.directed);
		json.add("strength", link.strength);
		link.idx = this.links.size();
		remoteObject.call("addLink", json);
		links.add(link);
	}
	
	public Link addLink(Node source, Node target, String label, boolean directed, float strength) {
		Link l = new Link(source, target, label, directed, strength);
		this.addLink(l);
		return l;
	}
	
	public void showGraph() {
		remoteObject.call("showGraph", new JsonObject());
	}
	
	public void restart() {
		remoteObject.call("restart", new JsonObject());
	}
	
	public void setCharge(float charge) {
		JsonObject params = new JsonObject();
		params.add("charge", charge);
		remoteObject.call("setCharge", params);
	}
	
	public void setLinkDistance(int distance) {
		JsonObject params = new JsonObject();
		params.add("linkdist", distance);
		remoteObject.call("setLinkDistance", params);
	}
	
	public void setFriction(int friction) {
		JsonObject params = new JsonObject();
		params.add("p", friction);
		remoteObject.call("setFriction", params);
	}
	
	public void setChargeDistance(int chargeDistance) {
		JsonObject params = new JsonObject();
		params.add("p", chargeDistance);
		remoteObject.call("setChargeDistance", params);
	}
	
	public void setGravity(float gravity) {
		JsonObject params = new JsonObject();
		params.add("p", gravity);
		remoteObject.call("setGravity", params);
	}
	
	public void removeAllNodes() {
		remoteObject.call("removeAllNodes", new JsonObject());
		this.nodes.clear();
		this.links.clear();
	}
	
}

